src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs

```diff
@@ -414,24 +415,33 @@ private void EmitObject(string objectFilePath, IReadOnlyCollection<DependencyNod
 
                 sectionWriter.EmitAlignment(nodeContents.Alignment);
 
+                bool hasInitialEntrypoint = false;
                 bool isMethod = node is IMethodBodyNode or AssemblyStubNode;
                 long thumbBit = _nodeFactory.Target.Architecture == TargetArchitecture.ARM && isMethod ? 1 : 0;
                 foreach (ISymbolDefinitionNode n in nodeContents.DefinedSymbols)
                 {
+                    hasInitialEntrypoint |= n.Offset == 0;
                     sectionWriter.EmitSymbolDefinition(
                         n == node ? currentSymbolName : GetMangledName(n),
                         n.Offset + thumbBit,
-                        n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0);
+                        n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0,
+                        altEntry: n.Offset != 0);
                     if (_nodeFactory.GetSymbolAlternateName(n) is string alternateName)
                     {
                         sectionWriter.EmitSymbolDefinition(
                             ExternCName(alternateName),
                             n.Offset + thumbBit,
                             n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0,
-                            global: true);
+                            global: true,
+                            altEntry: n.Offset != 0);
                     }
                 }
 
+                // Match the logic for KeepDataLayout in MachSection.CreateSection. If we are in a data section
+                // we don't need to have a symbol for beginning of the node data. For other nodes, particularly
+                // executable code, we enforce it though.
+                Debug.Assert(hasInitialEntrypoint || section.Type is SectionType.ReadOnly or SectionType.Writeable or SectionType.Uninitialized);
```

Comment:
Can we avoid duplicating the `section.Type is SectionType.ReadOnly or SectionType.Writeable or SectionType.Uninitialized` part?

Reply:
Done